pkg/controller/elasticache/user/setup.go

```diff
@@ -0,0 +1,31 @@
+package user
+
+import (
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	svcapitypes "github.com/crossplane/provider-aws/apis/elasticache/v1alpha1"
+)
+
+// SetupUser adds a controller that reconciles User.
+func SetupUser(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
```

Comment:
@namku I am guessing you are going to need to define additional functions for this type ([example](https://github.com/crossplane/provider-aws/blob/a8e175bc572c441188ee370a5023f6d726f68157/pkg/controller/rds/dbcluster/setup.go#L54)). Have you tested out using this resource?

Reply:
Hi @hasheddan and thanks for the review, yes I tried to create this resource with this steps (https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#local) and I could create the user, and delete too, but I going to review the example you share with me.

Reply:
I tried to add PreObserver function but when I add it and I try to create the User, the User is create but, all the time is spamming that message:
```
"message": "cannot create User in AWS: UserAlreadyExists: User exampleid already exists."
```
When I try to delete the User don't delete it from AWS and the message is:
```
"error": "User.elasticache.aws.crossplane.io\"userelasticache\" not found"
```
And if I try the same without that function, working properly, creating and deleting the user. can you give me some light about this?

the process I follow to try is:

1- Create cluster with kind.
2- Create ns crossplane-system
3- Create secret with aws credentials
4- Deploy providerconfig.yaml (AWS)
5- Execute Make generate
6- Deploy packages/crds
7- Execute "go run cmd/provider/main.go --debug"
8- Deploy example/elasticache/user.yaml
9- Delete example/elasticache/user.yaml

Reply:
@namku I think what you want to do here is first attempt to `Get` the user. If the user does not exist then you can `Create`. Likewise, with the delete operation, if you call `Delete` and the user no longer exists then there is no need to return an error because we have reached the desired state :)

Reply:
@hasheddan do we always need to override hooks for [external names](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#external-name)? Do we have some defaults for external names which could be just working here without overriding?

Reply:
@namku I believe we at least have to set a postObserve to indicate our [resource as ready](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#readiness-check), with something like:

```
func postObserve(_ context.Context, cr *v1alpha1.User, resp *elasticache.DescribeUsersOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
	if err != nil {
		return managed.ExternalObservation{}, err
	}
	if len(resp.Users) != 1 || resp.Users[0] == nil {
		return managed.ExternalObservation{}, errors.New("no user info found in response")
	}
	switch aws.StringValue(resp.Users[0].Status) {
	case "active":
		cr.SetConditions(xpv1.Available())
	case "modifying","deleting":
		cr.SetConditions(xpv1.Unavailable())
	}

	return obs, nil
}
```

Reply:
@turkenh @hasheddan fixed